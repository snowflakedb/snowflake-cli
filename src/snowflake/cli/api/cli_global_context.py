# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from contextlib import contextmanager
from contextvars import ContextVar
from dataclasses import dataclass, field, replace
from functools import wraps
from pathlib import Path
from typing import TYPE_CHECKING, Iterator, Optional

from snowflake.cli.api.connections import ConnectionContext, OpenConnectionCache
from snowflake.cli.api.exceptions import MissingConfiguration
from snowflake.cli.api.metrics import CLIMetrics
from snowflake.cli.api.output.formats import OutputFormat
from snowflake.cli.api.rendering.jinja import CONTEXT_KEY
from snowflake.connector import SnowflakeConnection

if TYPE_CHECKING:
    from snowflake.cli.api.project.definition_manager import DefinitionManager
    from snowflake.cli.api.project.schemas.project_definition import ProjectDefinition

_CONNECTION_CACHE = OpenConnectionCache()


@dataclass
class _CliGlobalContextManager:
    connection_context: ConnectionContext = field(default_factory=ConnectionContext)
    connection_cache: OpenConnectionCache = (
        _CONNECTION_CACHE  # by default, use global cache
    )

    output_format: OutputFormat = OutputFormat.TABLE
    silent: bool = False
    verbose: bool = False
    experimental: bool = False
    enable_tracebacks: bool = True

    metrics: CLIMetrics = field(default_factory=CLIMetrics)

    project_path_arg: str | None = None
    project_is_optional: bool = True
    project_env_overrides_args: dict[str, str] = field(default_factory=dict)

    # FIXME: this property only exists to help implement
    # nativeapp_definition_v2_to_v1 and single_app_and_package.
    # Consider changing the way this calculation is provided to commands
    # in order to remove this logic (then make project_definition a non-cloned @property)
    override_project_definition: ProjectDefinition | None = None

    _definition_manager: DefinitionManager | None = None

    # which properties invalidate our current DefinitionManager?
    DEFINITION_MANAGER_DEPENDENCIES = [
        "project_path_arg",
        "project_is_optional",
        "project_env_overrides_args",
    ]

    def reset(self):
        self.__init__()

    def clone(self) -> _CliGlobalContextManager:
        return replace(
            self,
            connection_context=self.connection_context.clone(),
            project_env_overrides_args=self.project_env_overrides_args.copy(),
            metrics=self.metrics.clone(),
        )

    def __setattr__(self, prop, val):
        if prop in self.DEFINITION_MANAGER_DEPENDENCIES:
            self._clear_definition_manager()

        super().__setattr__(prop, val)

    @property
    def project_definition(self) -> ProjectDefinition | None:
        if self.override_project_definition:
            return self.override_project_definition

        return self._definition_manager_or_raise().project_definition

    @property
    def project_root(self) -> Path:
        return Path(self._definition_manager_or_raise().project_root)

    @property
    def template_context(self) -> dict:
        return self._definition_manager_or_raise().template_context

    @property
    def connection(self) -> SnowflakeConnection:
        """
        Returns a connection for our configured context from the configured cache.
        By default, this is the global _CONNECTION_CACHE. If a matching connection
        does not already exist, creates a new connection and caches it.
        """
        self.connection_context.validate_and_complete()
        return self.connection_cache[self.connection_context]

    def _definition_manager_or_raise(self) -> DefinitionManager:
        """
        (Re-)parses project definition based on project args (project_path_arg and
        project_env_overrides_args). If we cannot provide a project definition
        (i.e. no snowflake.yml) and require one, raises MissingConfiguration.
        """
        from snowflake.cli.api.project.definition_manager import DefinitionManager

        # don't need to re-parse definition if we already have one
        if not self._definition_manager:
            dm = DefinitionManager(
                self.project_path_arg,
                {CONTEXT_KEY: {"env": self.project_env_overrides_args}},
            )
            if not dm.has_definition_file and not self.project_is_optional:
                raise MissingConfiguration(
                    "Cannot find project definition (snowflake.yml). Please provide a path to the project or run this command in a valid project directory."
                )
            self._definition_manager = dm

        return self._definition_manager

    def _clear_definition_manager(self):
        """
        Force re-calculation of definition_manager and its dependent attributes
        (template_context, project_definition, and project_root).
        """
        self._definition_manager = None


class _CliGlobalContextAccess:
    def __init__(self, manager: _CliGlobalContextManager):
        self._manager = manager

    @property
    def connection(self) -> SnowflakeConnection:
        return self._manager.connection

    @property
    def connection_context(self) -> ConnectionContext:
        return self._manager.connection_context

    @property
    def enable_tracebacks(self) -> bool:
        return self._manager.enable_tracebacks

    @property
    def metrics(self) -> CLIMetrics:
        return self._manager.metrics

    @property
    def output_format(self) -> OutputFormat:
        return self._manager.output_format

    @property
    def verbose(self) -> bool:
        return self._manager.verbose

    @property
    def experimental(self) -> bool:
        return self._manager.experimental

    @property
    def project_definition(self) -> ProjectDefinition | None:
        return self._manager.project_definition

    @property
    def project_root(self) -> Path | None:
        return self._manager.project_root

    @property
    def template_context(self) -> dict:
        return self._manager.template_context

    @property
    def silent(self) -> bool:
        if self._should_force_mute_intermediate_output:
            return True
        return self._manager.silent

    @property
    def _should_force_mute_intermediate_output(self) -> bool:
        """Computes whether cli_console output should be muted."""
        return self._manager.output_format == OutputFormat.JSON

    @property
    def snow_api_root(
        self,
    ) -> Optional[
        object
    ]:  # Should be Optional[Root], but we need local import for performance reasons
        from snowflake.core import Root

        if self.connection:
            return Root(self.connection)
        else:
            return None


_CLI_CONTEXT_MANAGER: ContextVar[_CliGlobalContextManager | None] = ContextVar(
    "cli_context", default=None
)


def get_cli_context_manager() -> _CliGlobalContextManager:
    mgr = _CLI_CONTEXT_MANAGER.get()
    if not mgr:
        mgr = _CliGlobalContextManager()
        _CLI_CONTEXT_MANAGER.set(mgr)
    return mgr


def get_cli_context() -> _CliGlobalContextAccess:
    return _CliGlobalContextAccess(get_cli_context_manager())


def span(span_name: str):
    """
    Decorator to start a command metrics span that encompasses a whole function

    Must be used instead of directly calling @get_cli_context().metrics.span(span_name)
    as a decorator to ensure that the cli context is grabbed at run time instead of at
    module load time, which would not reflect forking
    """

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            with get_cli_context().metrics.span(span_name):
                return func(*args, **kwargs)

        return wrapper

    return decorator


@contextmanager
def fork_cli_context(
    connection_overrides: dict | None = None,
    project_env_overrides: dict[str, str] | None = None,
    project_is_optional: bool | None = None,
    project_path: str | None = None,
) -> Iterator[_CliGlobalContextAccess]:
    """
    Forks the global CLI context, making changes that are only visible
    (e.g. via get_cli_context()) while inside this context manager.

    Please note that environment variable changes are only visible through
    the project definition; os.getenv / os.environ / get_env_value are not
    affected by these new values.
    """
    old_manager = get_cli_context_manager()
    new_manager = old_manager.clone()
    token = _CLI_CONTEXT_MANAGER.set(new_manager)

    if connection_overrides:
        new_manager.connection_context.update(**connection_overrides)

    if project_env_overrides:
        new_manager.project_env_overrides_args.update(project_env_overrides)

    if project_is_optional is not None:
        new_manager.project_is_optional = project_is_optional

    if project_path:
        new_manager.project_path_arg = project_path

    yield _CliGlobalContextAccess(new_manager)
    _CLI_CONTEXT_MANAGER.reset(token)
