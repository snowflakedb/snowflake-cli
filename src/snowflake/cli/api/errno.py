# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# General errors
NO_WAREHOUSE_SELECTED_IN_SESSION = 606
EMPTY_SQL_STATEMENT = 900

SQL_COMPILATION_ERROR = 1003
OBJECT_ALREADY_EXISTS_IN_DOMAIN = 1998
OBJECT_ALREADY_EXISTS = 2002
DOES_NOT_EXIST_OR_NOT_AUTHORIZED = 2003  # BASE_TABLE_OR_VIEW_NOT_FOUND
DUPLICATE_COLUMN_NAME = 2025
VIEW_EXPANSION_FAILED = 2037
DOES_NOT_EXIST_OR_CANNOT_BE_PERFORMED = (
    2043  # OBJECT_DOES_NOT_EXIST_OR_CANNOT_PERFORM_OPERATION
)
INSUFFICIENT_PRIVILEGES = 3001  # NOT_AUTHORIZED
INVALID_OBJECT_TYPE_FOR_SPECIFIED_PRIVILEGE = 3008
ROLE_NOT_ASSIGNED = 3013
NO_INDIVIDUAL_PRIVS = 3028
OBJECT_ALREADY_EXISTS_NO_PRIVILEGES = 3041

# Native Apps
APPLICATION_PACKAGE_MANIFEST_SPECIFIED_FILE_NOT_FOUND = 93003
APPLICATION_FILE_NOT_FOUND_ON_STAGE = 93009
CANNOT_GRANT_OBJECT_NOT_IN_APP_PACKAGE = 93011
CANNOT_GRANT_RESTRICTED_PRIVILEGE_TO_APP_PACKAGE_SHARE = 93012
APPLICATION_PACKAGE_VERSION_ALREADY_EXISTS = 93030
APPLICATION_PACKAGE_VERSION_NAME_TOO_LONG = 93035
APPLICATION_PACKAGE_PATCH_DOES_NOT_EXIST = 93036
APPLICATION_PACKAGE_MAX_VERSIONS_HIT = 93037
CANNOT_UPGRADE_FROM_LOOSE_FILES_TO_VERSION = 93044
CANNOT_UPGRADE_FROM_VERSION_TO_LOOSE_FILES = 93045
ONLY_SUPPORTED_ON_DEV_MODE_APPLICATIONS = 93046
NO_VERSIONS_AVAILABLE_FOR_ACCOUNT = 93054
NOT_SUPPORTED_ON_DEV_MODE_APPLICATIONS = 93055
APPLICATION_NO_LONGER_AVAILABLE = 93079
APPLICATION_INSTANCE_FAILED_TO_RUN_SETUP_SCRIPT = 93082
APPLICATION_INSTANCE_NO_ACTIVE_WAREHOUSE_FOR_CREATE_OR_UPGRADE = 93083
APPLICATION_INSTANCE_EMPTY_SETUP_SCRIPT = 93084
VERSION_REFERENCED_BY_RELEASE_DIRECTIVE = 93088
APPLICATION_PACKAGE_MANIFEST_CONTAINER_IMAGE_URL_BAD_VALUE = 93148
CANNOT_GRANT_NON_MANIFEST_PRIVILEGE = 93118
APPLICATION_OWNS_EXTERNAL_OBJECTS = 93128
APPLICATION_PACKAGE_PATCH_ALREADY_EXISTS = 93168
APPLICATION_PACKAGE_CANNOT_SET_EXTERNAL_DISTRIBUTION_WITH_SPCS = 93197
NATIVE_APPLICATION_MANIFEST_UNRECOGNIZED_FIELD = 93301
NATIVE_APPLICATION_MANIFEST_UNEXPECTED_VALUE_FOR_PROPERTY = 93302
NATIVE_APPLICATION_MANIFEST_GENERIC_JSON_ERROR = 93303
NATIVE_APPLICATION_MANIFEST_INVALID_SYNTAX = 93300
APPLICATION_REQUIRES_TELEMETRY_SHARING = 93321
CANNOT_DISABLE_MANDATORY_TELEMETRY = 93329
VERSION_NOT_ADDED_TO_RELEASE_CHANNEL = 512008
CANNOT_DISABLE_RELEASE_CHANNELS = 512001
RELEASE_DIRECTIVES_VERSION_PATCH_NOT_FOUND = 93036
RELEASE_DIRECTIVE_DOES_NOT_EXIST = 93090
VERSION_DOES_NOT_EXIST = 93031
VERSION_NOT_IN_RELEASE_CHANNEL = 512010
ACCOUNT_DOES_NOT_EXIST = 1999
ACCOUNT_HAS_TOO_MANY_QUALIFIERS = 906
CANNOT_MODIFY_RELEASE_CHANNEL_ACCOUNTS = 512017
VERSION_ALREADY_ADDED_TO_RELEASE_CHANNEL = 512005
MAX_VERSIONS_IN_RELEASE_CHANNEL_REACHED = 512004
MAX_UNBOUND_VERSIONS_REACHED = 512023
CANNOT_DEREGISTER_VERSION_ASSOCIATED_WITH_CHANNEL = 512021
TARGET_ACCOUNT_USED_BY_OTHER_RELEASE_DIRECTIVE = 93091


ERR_JAVASCRIPT_EXECUTION = 100132

SNOWSERVICES_IMAGE_REPOSITORY_IMAGE_IMPORT_TO_NATIVE_APP_FAIL = 397007
SNOWSERVICES_IMAGE_MANIFEST_NOT_FOUND = 397012
SNOWSERVICES_IMAGE_REPOSITORY_FAILS_TO_RETRIEVE_IMAGE_HASH_NEW = 397013

NO_REFERENCE_SET_FOR_DEFINITION = 505019
NO_ACTIVE_REF_DEFINITION_WITH_REF_NAME_IN_APPLICATION = 505026
