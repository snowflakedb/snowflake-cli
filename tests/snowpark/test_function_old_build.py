# Copyright (c) 2024 Snowflake Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
from pathlib import Path
from textwrap import dedent
from unittest import mock

import pytest
from snowflake.cli.api.errno import DOES_NOT_EXIST_OR_NOT_AUTHORIZED
from snowflake.connector import ProgrammingError

from tests_common import IS_WINDOWS

if IS_WINDOWS:
    pytest.skip("Requires further refactor to work on Windows", allow_module_level=True)


mock_session_has_warehouse = mock.patch(
    "snowflake.cli.api.sql_execution.SqlExecutionMixin.session_has_warehouse",
    lambda _: True,
)


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager")
@mock_session_has_warehouse
def test_deploy_function(
    mock_object_manager,
    mock_connector,
    mock_ctx,
    runner,
    project_directory,
):
    mock_object_manager.return_value.describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    with project_directory("snowpark_functions") as project_dir:
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
            catch_exceptions=False,
        )

    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(project_dir).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace function IDENTIFIER('MockDatabase.MockSchema.func1')(a string default 'default value', b variant)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.func1_handler'
            packages=()
            """
        ).strip(),
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager")
@mock_session_has_warehouse
def test_deploy_function_with_external_access(
    mock_object_manager,
    mock_connector,
    mock_ctx,
    runner,
    project_directory,
):
    mock_object_manager.return_value.show.return_value = [
        {"name": "external_1", "type": "EXTERNAL_ACCESS"},
        {"name": "external_2", "type": "EXTERNAL_ACCESS"},
    ]
    mock_object_manager.return_value.describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    ctx = mock_ctx()
    mock_connector.return_value = ctx

    with project_directory("snowpark_function_external_access") as project_dir:
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
            catch_exceptions=False,
        )

    assert result.exit_code == 0, result.output
    assert ctx.get_queries() == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(project_dir).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace function IDENTIFIER('MockDatabase.MockSchema.func1')(a string, b variant)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.func1_handler'
            packages=()
            external_access_integrations=(external_1, external_2)
            secrets=('cred'=cred_name, 'other'=other_name)
            """
        ).strip(),
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager")
@mock_session_has_warehouse
def test_deploy_function_secrets_without_external_access(
    mock_object_manager,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    os_agnostic_snapshot,
):
    mock_object_manager.return_value.show.return_value = [
        {"name": "external_1", "type": "EXTERNAL_ACCESS"},
        {"name": "external_2", "type": "EXTERNAL_ACCESS"},
    ]
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_function_secrets_without_external_access"):
        result = runner.invoke(
            [
                "snowpark",
                "deploy",
            ],
        )

    assert result.exit_code == 1, result.output
    assert result.output == os_agnostic_snapshot


@mock.patch("snowflake.connector.connect")
@mock_session_has_warehouse
def test_deploy_function_no_changes(
    mock_connector,
    runner,
    mock_ctx,
    mock_cursor,
    project_directory,
):
    rows = [
        ("packages", '["foo==1.2.3", "bar>=3.0.0"]'),
        ("handler", "app.func1_handler"),
        ("returns", "string"),
        ("imports", "dev_deployment/my_snowpark_project/app.py"),
        ("runtime_version", "3.10"),
    ]

    queries, result, project_dir = _deploy_function(
        rows,
        mock_connector,
        runner,
        mock_ctx,
        mock_cursor,
        project_directory,
        "--replace",
    )

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.func1(a string default 'default value', b variant)",
            "status": "packages updated",
            "type": "function",
        }
    ]
    assert queries == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(project_dir).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/ auto_compress=false parallel=4 overwrite=True",
    ]


@mock.patch("snowflake.connector.connect")
@mock_session_has_warehouse
def test_deploy_function_needs_update_because_packages_changes(
    mock_connector,
    runner,
    mock_ctx,
    mock_cursor,
    project_directory,
):
    rows = [
        ("packages", '["foo==1.2.3"]'),
        ("handler", "main.py:app"),
        ("returns", "table(variant)"),
    ]

    queries, result, project_dir = _deploy_function(
        rows,
        mock_connector,
        runner,
        mock_ctx,
        mock_cursor,
        project_directory,
        "--replace",
    )

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.func1(a string default 'default value', b variant)",
            "status": "definition updated",
            "type": "function",
        }
    ]
    assert queries == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(project_dir).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/ auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace function IDENTIFIER('MockDatabase.MockSchema.func1')(a string default 'default value', b variant)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.func1_handler'
            packages=('foo==1.2.3','bar>=3.0.0')
            """
        ).strip(),
    ]


@mock.patch("snowflake.connector.connect")
@mock_session_has_warehouse
def test_deploy_function_needs_update_because_handler_changes(
    mock_connector,
    runner,
    mock_ctx,
    mock_cursor,
    project_directory,
):
    rows = [
        ("packages", '["foo==1.2.3", "bar>=3.0.0"]'),
        ("handler", "main.py:oldApp"),
        ("returns", "table(variant)"),
    ]

    queries, result, project_dir = _deploy_function(
        rows,
        mock_connector,
        runner,
        mock_ctx,
        mock_cursor,
        project_directory,
        "--replace",
    )

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": "MockDatabase.MockSchema.func1(a string default 'default value', b variant)",
            "status": "definition updated",
            "type": "function",
        }
    ]
    assert queries == [
        "create stage if not exists IDENTIFIER('MockDatabase.MockSchema.dev_deployment') comment='deployments managed by Snowflake CLI'",
        f"put file://{Path(project_dir).resolve()}/app.py @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/"
        f" auto_compress=false parallel=4 overwrite=True",
        dedent(
            """\
            create or replace function IDENTIFIER('MockDatabase.MockSchema.func1')(a string default 'default value', b variant)
            copy grants
            returns string
            language python
            runtime_version=3.10
            imports=('@MockDatabase.MockSchema.dev_deployment/my_snowpark_project/app.py')
            handler='app.func1_handler'
            packages=('foo==1.2.3','bar>=3.0.0')
            """
        ).strip(),
    ]


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_function_fully_qualified_name_duplicated_database(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    os_agnostic_snapshot,
):
    number_of_functions_in_project = 6
    mock_om_describe.side_effect = [
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ] * number_of_functions_in_project
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_function_fully_qualified_name") as tmp_dir:
        result = runner.invoke(["snowpark", "deploy"])
        assert result.output == os_agnostic_snapshot(name="database error")


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_function_fully_qualified_name_duplicated_schema(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    os_agnostic_snapshot,
):
    number_of_functions_in_project = 6
    mock_om_describe.side_effect = [
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ] * number_of_functions_in_project
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_function_fully_qualified_name") as tmp_dir:
        alter_snowflake_yml(
            tmp_dir / "snowflake.yml",
            parameter_path="snowpark.functions.5.name",
            value="custom_schema.fqn_function_error",
        )
        result = runner.invoke(["snowpark", "deploy"])
        assert result.output == os_agnostic_snapshot(name="schema error")


@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.describe")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager.show")
@mock_session_has_warehouse
def test_deploy_function_fully_qualified_name(
    mock_om_show,
    mock_om_describe,
    mock_conn,
    runner,
    mock_ctx,
    project_directory,
    alter_snowflake_yml,
    os_agnostic_snapshot,
):
    number_of_functions_in_project = 6
    mock_om_describe.side_effect = [
        ProgrammingError(errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED),
    ] * number_of_functions_in_project
    ctx = mock_ctx()
    mock_conn.return_value = ctx

    with project_directory("snowpark_function_fully_qualified_name") as tmp_dir:
        alter_snowflake_yml(
            tmp_dir / "snowflake.yml",
            parameter_path="snowpark.functions.5.name",
            value="fqn_function3",
        )
        result = runner.invoke(["snowpark", "deploy"])
        assert result.exit_code == 0
        assert result.output == os_agnostic_snapshot(name="ok")


@pytest.mark.parametrize(
    "parameter_type,default_value",
    [
        ("string", None),
        ("string", ""),
        ("int", None),
        ("variant", None),
        ("bool", None),
    ],
)
@mock.patch("snowflake.connector.connect")
@mock.patch("snowflake.cli._plugins.snowpark.commands.ObjectManager")
@mock_session_has_warehouse
def test_deploy_function_with_empty_default_value(
    mock_object_manager,
    mock_connector,
    mock_ctx,
    runner,
    project_directory,
    alter_snowflake_yml,
    parameter_type,
    default_value,
):
    mock_object_manager.return_value.describe.side_effect = ProgrammingError(
        errno=DOES_NOT_EXIST_OR_NOT_AUTHORIZED
    )
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    with project_directory("snowpark_functions") as project_dir:
        snowflake_yml = project_dir / "snowflake.yml"
        for param, value in [("type", parameter_type), ("default", default_value)]:
            alter_snowflake_yml(
                snowflake_yml,
                parameter_path=f"snowpark.functions.0.signature.0.{param}",
                value=value,
            )
        alter_snowflake_yml(
            snowflake_yml,
            parameter_path=f"snowpark.functions.0.runtime",
            value="3.10",
        )
        result = runner.invoke(
            ["snowpark", "deploy", "--format", "json"], catch_exceptions=False
        )
    default_value_json = default_value
    if default_value is None:
        default_value_json = "null"
    elif parameter_type == "string":
        default_value_json = f"'{default_value}'"

    assert result.exit_code == 0, result.output
    assert json.loads(result.output) == [
        {
            "object": f"MockDatabase.MockSchema.func1(a {parameter_type} default {default_value_json}, b variant)",
            "status": "created",
            "type": "function",
        }
    ]


@mock.patch("snowflake.connector.connect")
def test_execute_function(mock_connector, runner, mock_ctx):
    ctx = mock_ctx()
    mock_connector.return_value = ctx
    result = runner.invoke(
        [
            "snowpark",
            "execute",
            "function",
            "functionName(42, 'string')",
        ]
    )

    assert result.exit_code == 0, result.output
    assert ctx.get_query() == "select functionName(42, 'string')"


def _deploy_function(
    rows,
    mock_connector,
    runner,
    mock_ctx,
    mock_cursor,
    project_directory,
    *args,
):
    ctx = mock_ctx(mock_cursor(rows=rows, columns=[]))
    mock_connector.return_value = ctx
    with (
        mock.patch(
            "snowflake.cli._plugins.snowpark.commands.ObjectManager.describe"
        ) as om_describe,
        mock.patch(
            "snowflake.cli._plugins.snowpark.commands.ObjectManager.show"
        ) as om_show,
    ):
        om_describe.return_value = rows

        with project_directory("snowpark_functions") as temp_dir:
            (Path(temp_dir) / "requirements.snowflake.txt").write_text(
                "foo==1.2.3\nbar>=3.0.0"
            )
            result = runner.invoke(
                [
                    "snowpark",
                    "deploy",
                    "--format",
                    "json",
                    *args,
                ]
            )
    queries = ctx.get_queries()
    return queries, result, temp_dir
