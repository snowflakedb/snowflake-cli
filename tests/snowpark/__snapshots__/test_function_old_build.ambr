# serializer version: 1
# name: test_deploy_function_fully_qualified_name[ok]
  '''
  Performing initial validation
  Checking remote state
  Preparing required stages and artifacts
    Creating (if not exists) stage: dev_deployment
    Uploading app.zip to @MockDatabase.MockSchema.dev_deployment/my_snowpark_project/
  Creating Snowpark entities
    Creating function custom_db.custom_schema.fqn_function
    Creating function custom_schema.fqn_function_only_schema
    Creating function custom_schema.schema_function
    Creating function custom_db.PUBLIC.database_function
    Creating function custom_db.custom_schema.database_function
    Creating function custom_database.custom_schema.fqn_function3
  +------------------------------------------------------------------------------+
  | object                                                  | type     | status  |
  |---------------------------------------------------------+----------+---------|
  | custom_db.custom_schema.fqn_function(name string)       | function | created |
  | MockDatabase.custom_schema.fqn_function_only_schema(nam | function | created |
  | e string)                                               |          |         |
  | MockDatabase.custom_schema.schema_function(name string) | function | created |
  | custom_db.MockSchema.database_function(name string)     | function | created |
  | custom_db.custom_schema.database_function(name string)  | function | created |
  | custom_database.custom_schema.fqn_function3(name        | function | created |
  | string)                                                 |          |         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_fully_qualified_name_duplicated_database[database error]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Database provided but name                                                   |
  | 'custom_database.custom_schema.fqn_function_error' is fully qualified name.  |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_fully_qualified_name_duplicated_schema[schema error]
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | Schema provided but name 'custom_schema.fqn_function_error' is fully         |
  | qualified name.                                                              |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_deploy_function_secrets_without_external_access
  '''
  Performing initial validation
  Checking remote state
  +- Error ----------------------------------------------------------------------+
  | func1 defined with secrets but without external integration.                 |
  +------------------------------------------------------------------------------+
  
  '''
# ---
